flocker/node/agents/functional/test_gce.py

```diff
@@ -322,13 +388,91 @@ def test_attach_elsewhere_attached_volume(self):
             attach_to=api.compute_instance_id(),
         )
 
+    def test_duplicated_calls(self):
+        """
+        Verify that if every call to the :class:`GCEAtomicOperation` s is
+        duplicated that we handle the errors correctly.
+
+        This should force some specific scheduling situations that resemble
+        race conditions with another agent trying to converge to the same
+        state, or a condition where the dataset agent as rebooted after a crash
+        that happened in the middle of an :class:`IBlockDeviceAPI` call.
+
+        In these situations we should verify that the second call to many of
+        the underlying atomic methods would result in the correct underlying
+        :class:`VolumeException`.
+        """
+        actual_api = gceblockdeviceapi_for_test(self)
+        volume_manager = actual_api._volume_manager
+        api = actual_api.set(
+            '_volume_manager',
+            repeat_call_proxy_for(IGCEVolumeManager, volume_manager)
+        )
+
+        dataset_id = uuid4()
+
+        # There is no :class:`VolumeException` for creating an already created
+        # volume. Thus, GCE just raises its own custom exception in that case.
+        self.assertThat(
+            lambda: api.create_volume(
+                dataset_id=dataset_id,
+                size=get_minimum_allocatable_size()
+            ),
+            Raises(MatchesException(GCEVolumeException))
+        )
+
+        volumes = api.list_volumes()
+
+        self.assertThat(
+            volumes,
+            AnyMatch(MatchesStructure(dataset_id=Equals(dataset_id)))
+        )
+        volume = next(v for v in volumes if v.dataset_id == dataset_id)
+
+        compute_instance_id = api.compute_instance_id()
+
+        self.assertThat(
+            lambda: api.attach_volume(
+                blockdevice_id=volume.blockdevice_id,
+                attach_to=compute_instance_id,
+            ),
+            Raises(MatchesException(AlreadyAttachedVolume))
+        )
+
+        self.assertThat(
+            api.get_device_path(volume.blockdevice_id).path,
+            Contains('/dev/sd')
+        )
+
+        api.detach_volume(
+            blockdevice_id=volume.blockdevice_id,
+        )
```

Comment:
:grey_question: possibly add a comment that detach_volume isn't in the interface.  Not sure what our commenting guidelines are but I had to double check to make sure this wasn't calling it 4 times.